File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    private List<String> result;
    private boolean[] visited;
    
    public String[] solution(String[][] tickets) {
        result = new ArrayList<>();
        visited = new boolean[tickets.length];
        String[] answer = {};
        dfs("ICN", "ICN", tickets, 0);
        Collections.sort(result);
        answer = result.get(0).split(",");
        return answer;
    }
    
    private void dfs(String start, String root, String[][] tickets, int depth) {
        if (depth == tickets.length) {
            result.add(root);
            return;
        }
        
        for (int i = 0; i < tickets.length; i++) {
            if (Objects.equals(start,tickets[i][0]) && !visited[i]) {
                visited[i] = true;
                dfs(tickets[i][1], root + "," + tickets[i][1], tickets, depth + 1);
                visited[i] = false;
            }
        }
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict
result = []
graph = defaultdict(list)

def dfs(s):
    while graph[s]:
        
        dfs(graph[s].pop(0))
        
    if not graph[s]:
        result.append(s)
        return

def solution(tickets):
    for a,b in tickets:
        graph[a].append(b)
    for a in graph:
        graph[a].sort()
        
    dfs("ICN")
    return result[::-1]